lib/cocoapods/installer/analyzer/target_generator.rb

```diff
@@ -0,0 +1,397 @@
+          end
+          aggregate_targets
+        end
+
+
+
+        # Setup the pod targets for an aggregate target. Deduplicates resulting
+        # targets by grouping by platform and subspec by their root
+        # to create a {PodTarget} for each spec.
+        #
+        # @param  [Hash{TargetDefinition => Array<ResolverSpecification>}] resolver_specs_by_target
+        #         the resolved specifications grouped by their target definition.
+        #
+        # @param  [Hash{TargetDefinition => Array<TargetInspectionResult>}] target_inspections
+        #         the user target inspections used to construct the aggregate and pod targets.
+        #
+        # @param  [Sandbox] sandbox
+        #         the sandbox into which the pod targets will be installed
+        #
+        # @param  [InstallationOptions] installation_options
+        #         the installation options to use when generating the targets
+        #
+        # @return [Array<PodTarget>]
+        #
+        def self.generate_pod_targets(resolver_specs_by_target, target_inspections, sandbox, installation_options)
+          if installation_options.deduplicate_targets?
+            distinct_targets = resolver_specs_by_target.each_with_object({}) do |(target_definition, dependent_specs), hash|
+              dependent_specs.group_by(&:root).each do |root_spec, resolver_specs|
+                all_specs = resolver_specs.map(&:spec)
+                test_specs, specs = all_specs.partition(&:test_specification?)
+                pod_variant = PodVariant.new(specs, test_specs, target_definition.platform, target_definition.uses_frameworks?)
+                hash[root_spec] ||= {}
+                (hash[root_spec][pod_variant] ||= []) << target_definition
+                hash[root_spec].keys.find { |k| k == pod_variant }.test_specs.concat(test_specs).uniq!
+              end
+            end
+
+            pod_targets = distinct_targets.flat_map do |_root, target_definitions_by_variant|
+              suffixes = PodVariantSet.new(target_definitions_by_variant.keys).scope_suffixes
+              target_definitions_by_variant.flat_map do |variant, target_definitions|
+                generate_pod_target(target_definitions, target_inspections, variant.specs + variant.test_specs, sandbox, installation_options, :scope_suffix => suffixes[variant])
+              end
+            end
+          else
+            dedupe_cache = {}
+            pod_targets = resolver_specs_by_target.flat_map do |target_definition, specs|
+              grouped_specs = specs.group_by(&:root).values.uniq
+              grouped_specs.flat_map do |pod_specs|
+                generate_pod_target([target_definition], target_inspections, pod_specs.map(&:spec), sandbox, installation_options).scoped(dedupe_cache)
+              end
+            end
+          end
+
+          all_resolver_specs = resolver_specs_by_target.values.flatten.map(&:spec).uniq
```

Comment:
The code starting from here was duplicated within each branch of the `if`, with the only difference being:

```
if installation_options.integrate_targets?
  ...
  # within target.test_dependent_targets_by_spec_name =
  hash[test_spec.name] = filter_dependencies(test_dependencies, pod_targets_by_name, target
else
  ...
 # within target.test_dependent_targets_by_spec_name =
  hash[test_spec.name] = pod_targets.reject { |t| test_dependencies[t.root_spec].nil? }
end
```

now both branches are using `filter_dependencies`, but is that valid? tests were passing but not sure if there's an edge case not covered by the tests

Reply:
@segiddins this is the only thing that could potential affect functionality as far as I can tell